src/common/src/types/ordered.rs

```diff
@@ -0,0 +1,174 @@
+
+use std::cmp::{Ord, Ordering};
+use std::ops::Deref;
+
+use paste::paste;
+use risingwave_common_proc_macro::EstimateSize;
+
+use crate::estimate_size::EstimateSize;
+use crate::for_all_scalar_variants;
+use crate::types::{Datum, DatumRef, ScalarImpl, ScalarRefImpl};
+use crate::util::sort_util::{cmp_datum, partial_cmp_datum, OrderType};
+
+macro_rules! gen_default_partial_cmp_scalar_ref_impl {
+    ($( { $variant_name:ident, $suffix_name:ident, $scalar:ty, $scalar_ref:ty } ),*) => {
+        pub fn default_partial_cmp_scalar_ref_impl(lhs: ScalarRefImpl<'_>, rhs: ScalarRefImpl<'_>) -> Option<Ordering> {
+            match (lhs, rhs) {
+                $((ScalarRefImpl::$variant_name(lhs_inner), ScalarRefImpl::$variant_name(ref rhs_inner)) => Some(lhs_inner.cmp(rhs_inner)),)*
+                _ => None,
+            }
+        }
+    };
+}
+for_all_scalar_variants!(gen_default_partial_cmp_scalar_ref_impl);
+
+/// Variant of [`PartialOrd`] that compares with default order.
+pub trait DefaultPartialOrd {
+    fn default_partial_cmp(&self, other: &Self) -> Option<Ordering>;
+}
+
+/// Variant of [`Ord`] that compares with default order.
+pub trait DefaultOrd: DefaultPartialOrd {
+    fn default_cmp(&self, other: &Self) -> Ordering;
+}
+
+impl DefaultPartialOrd for ScalarImpl {
+    fn default_partial_cmp(&self, other: &ScalarImpl) -> Option<Ordering> {
+        self.as_scalar_ref_impl()
+            .default_partial_cmp(&other.as_scalar_ref_impl())
+    }
+}
+
+impl DefaultOrd for ScalarImpl {
+    fn default_cmp(&self, other: &ScalarImpl) -> Ordering {
+        self.as_scalar_ref_impl()
+            .default_cmp(&other.as_scalar_ref_impl())
+    }
+}
+
+impl DefaultPartialOrd for ScalarRefImpl<'_> {
+    fn default_partial_cmp(&self, other: &ScalarRefImpl<'_>) -> Option<Ordering> {
+        default_partial_cmp_scalar_ref_impl(*self, *other)
+    }
+}
+
+impl DefaultOrd for ScalarRefImpl<'_> {
+    fn default_cmp(&self, other: &ScalarRefImpl<'_>) -> Ordering {
+        self.default_partial_cmp(other)
+            .unwrap_or_else(|| panic!("cannot compare {self:?} with {other:?}"))
+    }
+}
+
+impl DefaultPartialOrd for Datum {
+    fn default_partial_cmp(&self, other: &Datum) -> Option<Ordering> {
+        partial_cmp_datum(self, other, OrderType::default())
+    }
+}
+
+impl DefaultOrd for Datum {
+    fn default_cmp(&self, other: &Datum) -> Ordering {
+        cmp_datum(self, other, OrderType::default())
+    }
+}
+
+impl DefaultPartialOrd for DatumRef<'_> {
+    fn default_partial_cmp(&self, other: &DatumRef<'_>) -> Option<Ordering> {
+        partial_cmp_datum(*self, *other, OrderType::default())
+    }
+}
+
+impl DefaultOrd for DatumRef<'_> {
+    fn default_cmp(&self, other: &DatumRef<'_>) -> Ordering {
+        cmp_datum(*self, *other, OrderType::default())
+    }
+}
+
+/// Wrapper of [`ScalarImpl`] that can be sorted in default order.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, EstimateSize)]
+pub struct OrdScalarImpl {
```

Comment:
Maybe we can have a generic wrapper of `DefaultOrdered`, whose functionality is to delegate `std::cmp::Ord` to `DefaultOrd`. 🤩

```rust
trait Row {}

trait DefaultOrd {
    fn cmp_default(&self, _: &Self) -> std::cmp::Ordering;
}

struct DefaultOrdered<T>(T);

impl<T: DefaultOrd> PartialEq for DefaultOrdered<T> {
    fn eq(&self, _other: &Self) -> bool {
        todo!()
    }
}

impl<T: DefaultOrd> PartialOrd for DefaultOrdered<T> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        self.0.cmp_default(&other.0).into()
    }
}

impl<T: DefaultOrd> Eq for DefaultOrdered<T> {}

impl<T: DefaultOrd> Ord for DefaultOrdered<T> {
    fn cmp(&self, _: &Self) -> std::cmp::Ordering {
        std::cmp::Ordering::Equal
    }
}

impl<R: Row> Row for DefaultOrdered<R> {}

type DefaultOrdOwnedRow = DefaultOrdered<OwnedRow>;
type DefaultOrdScalarImpl = DefaultOrdered<ScalarImpl>;
```